Content.Server/Fluids/EntitySystems/SpraySystem.cs

```diff
@@ -0,0 +1,139 @@
+
+namespace Content.Server.Fluids.EntitySystems;
+
+public class SpraySystem : EntitySystem
+{
+    [Dependency] private readonly ActionBlockerSystem _actionBlockerSystem = default!;
+    [Dependency] private readonly PopupSystem _popupSystem = default!;
+    [Dependency] private readonly SolutionContainerSystem _solutionContainerSystem = default!;
+    [Dependency] private readonly VaporSystem _vaporSystem = default!;
+    [Dependency] private readonly IGameTiming _gameTiming = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<SprayComponent, AfterInteractEvent>(OnAfterInteract, after: new []{ typeof(FireExtinguisherSystem) });
+    }
+
+    private void OnAfterInteract(EntityUid uid, SprayComponent component, AfterInteractEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        args.Handled = true;
+        if (!_actionBlockerSystem.CanInteract(args.User))
+            return;
+
+        if (!_solutionContainerSystem.TryGetSolution(uid, SprayComponent.SolutionName, out var solution))
+            return;
+
+        var ev = new SprayAttemptEvent(args.User);
+        RaiseLocalEvent(uid, ev, false);
+        if (ev.Cancelled)
+            return;
+
+        var curTime = _gameTiming.CurTime;
+        if (TryComp<ItemCooldownComponent>(uid, out var cooldown)
+            && curTime < cooldown.CooldownEnd)
+            return;
+
+        if (solution.CurrentVolume <= 0)
+        {
+            _popupSystem.PopupEntity( Loc.GetString("spray-component-is-empty-message"),uid,
+                Filter.Entities(args.User));
+            return;
+        }
+
+        var playerPos = Transform(args.User).Coordinates;
+
+        if (args.ClickLocation.GetGridId(EntityManager) != playerPos.GetGridId(EntityManager))
+            return;
+
+        var direction = (args.ClickLocation.Position - playerPos.Position).Normalized;
+        var threeQuarters = direction * 0.75f;
+        var quarter = direction * 0.25f;
+
+        var amount = Math.Max(Math.Min((solution.CurrentVolume / component.TransferAmount).Int(), component.VaporAmount), 1);
+
+        var spread = component.VaporSpread / amount;
+
+        for (var i = 0; i < amount; i++)
+        {
+            var rotation = new Angle(direction.ToAngle() + Angle.FromDegrees(spread * i) -
+                                     Angle.FromDegrees(spread * (amount - 1) / 2));
+
+            var (_, diffPos) = args.ClickLocation - playerPos;
+            var diffNorm = diffPos.Normalized;
+            var diffLength = diffPos.Length;
+
+            var target = Transform(args.User).Coordinates
+                .Offset((diffNorm + rotation.ToVec()).Normalized * diffLength + quarter);
+
+            if (target.TryDistance(EntityManager, playerPos, out var distance) && distance > component.SprayDistance)
+                target = Transform(args.User).Coordinates
+                    .Offset(diffNorm * component.SprayDistance);
```

Comment:
I'm fine with ECSing this and merging it but jeez this needs refactoring. It looks like it's still based on gridcoordinates so I imagine it'd be jank if you're buckled.

Reply:
yeah its quite jank in general